varname.py

```diff
@@ -461,16 +484,88 @@ def __repr__(self) -> str:
         return (f"<{self.__class__.__name__} "
                 f"(name={self.name!r}, value={self.value!r})>")
 
-def _get_frame(caller: int) -> FrameType:
-    """Get the frame at `caller` depth"""
+def _debug(msg: str, frame: Optional[FrameType] = None) -> None:
+    """Print the debug message"""
+    if not DEBUG:
+        return
+    if frame is not None:
+        frameinfo = inspect.getframeinfo(frame)
+        msg = (f'{msg} [In {frameinfo.function!r} at '
+               f'{frameinfo.filename}:{frameinfo.lineno}]')
+    sys.stderr.write(f'[{__name__}] DEBUG: {msg}\n')
+
+def _get_executing(
+        caller: int,
+        ignore: Optional[
+            List[Union[ModuleType, Tuple[ModuleType, str]]]
+        ] = None
+) -> FrameType:
+    """This function makes sure that we get the desired frame,
+    by caller and ignroe.
+
+    The caller is the desired stack we want after excluding the frames we want
+    to ignore. By default, this module (varname) and the builtin modules will
+    be ignored.
+
+    The frame_index defaults to 1 to exclude the calls from inside this module.
+    """
+    # Use loop instead of recursion to avoid creating additional stacks
     try:
-        return sys._getframe(caller + 1)
+        # We are at least skipping 2 frames:
+        # one is this function, the other is any varname API that uses this
+        # Don't bother to test them, just simply skip.
+        frame_index = 2
+
+        while caller > 0:
+            frame = sys._getframe(frame_index)
+            # loot at next frame anyway at next iteration
+            frame_index += 1
+            module = inspect.getmodule(frame)
+            exect = executing.Source.executing(frame)
+
+            if module is sys.modules[__name__]:
+                _debug('Skipping frame from varname', frame)
+                continue
+
+            if module and module.__name__ in sys.builtin_module_names:
+                # havn't find a way to compose a test for this, skip for now
+                continue # pragma: no cover
+
+            if ignore and (
+                    module in ignore or
+                    (module, exect.code_qualname()) in ignore
+            ):
+                _debug('Ignored', frame)
+                continue
+
+            if ignore and '.' in module.__name__:
+                # if asyncio specified, asyncio.runners, asyncio.events, etc
+                # should be all ignored
+                modnames = module.__name__.split('.')[:-1]
+                if any(sys.modules['.'.join(modnames[:i+1])] in ignore
+                       for i, _ in enumerate(modnames)):
+
+                    _debug('Ignored', frame)
+                    continue
```

Comment:
What about specifying `(asyncio, 'function_qualname')` where the function is actually in `asyncio.runners`? This prefix checking here only works for ignoring entire modules.

Reply:
Then I kind of have to scan all qualnames in `asyncio.runners` to verify `function_qualname` is from it?

Reply:
Suppose `module.__name__ == 'foo.bar.spam' and exect.code_qualname() == 'func'`. Then we need to check `(foo.bar.spam, 'func') in ignore or (foo.bar, 'func') in ignore or (foo, 'func') in ignore`.

Reply:
Then what if there is a `func` in `foo.bar` or `foo` that one doesn't want to ignore?

Reply:
Currently your code will ignore any function in `foo.bar.spam` if I set `ignore=[foo]`. I'm not entirely sure if this is a good idea but I can see the advantage so let's assume we keep that.

Also currently your code will not ignore the function `func` in `foo.bar.spam` if I set `ignore=[(foo, 'func')]`. I think that's inconsistent and confusing. You have this nice prefix wildcard system for entire modules but not functions.

Reply:
I assume that if a pair (module and qualname) is received, one is ignoring an exact call (which can be located with the module (not the parent module) and qualname exactly). Otherwise, if a module is passed, I assume one wants to do some fuzzy ignoring (such as any calls from that module and submodules).

Reply:
I was actually thinking about ignoring all frames from python standard libraries, instead of just modules from `sys.builtin_module_names`, since developers barely have a chance to touch them. However, I didn't find a good way to list all standard libraries. Most of the solutions are trying to walk through the library path. 

In this way, developers don't have to worried about calls from libraries, such as `typing` if varname retrieving gets type annotations involved, `asyncio` if the retrieving is in async context, etc

Reply:
But this is somehow too aggresive.